### DISCUSSION TITLE
Notes about default styles

### DISCUSSION THREAD

**Discussion by pi0** [unjs/eslint-config#3] [General]
## Preface

JavaScript is a lean language. Unlike most languages, you have more than one choice to do the same thing. While this is a powerful side of JavaScript, it also makes inconsistencies and arguments across larger code bases. While most languages have a unified code style guide and solid spec including [Python PEP8](https://peps.python.org/pep-0008/), [PHP PSR-1](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-1-basic-coding-standard.md), and [GoLang](https://github.com/uber-go/guide/blob/master/style.md) and [Rust](https://github.com/rust-lang/fmt-rfcs), Javascript has many variations. [Standard JS](https://standardjs.com/) and [Prettier](https://prettier.io/) are probably the most widely adopted answers to it and endorsed by large communities.

Code style is a delicate matter for every single developer. In this Article, I tried to write the story and reasonings behind the new code style and the switch to [Prettier](https://prettier.io/) standard. What you are reading in the rest of the article, is entirely my POV and in no way endorses the recommendations for any ecosystem or community.

## StandardJS

[Standard JS](https://standardjs.com/) is one of the oldest code styles for javascript. I had been advocating [Standard JS](https://standardjs.com/) code style for **over 5+ years** after switching from [XO](https://github.com/xojs/xo) for all of my javascript codebases. I advocated the Standard style for [Nuxt Community](https://github.com/nuxt-community) which finally became our standard defacto and recommended style for Nuxt project via [nuxt/eslint-config](https://github.com/nuxt/eslint-config).

Standard JS provides both code styles, and also linting guidelines. You can see [all rules here](https://standardjs.com/rules.html). Here is a summary of important style specs:

- ..2 space indentation
- Single quotes for 'strings'
- No trailing commas
- No semicolons (unless needed)

If I want to describe the above rules, they are **minimal**, strict enough, and readable.

Over time many popular projects diverged from StandardJS and chose their custom rules. Including many of the ones currently highlighted on the landing page!

StandardJS is mainly adopted via ESLint formatter even tough it has a dedicated CLI (npm downloads: 2.5M for eslint config vs 300K for standard cli).  Meanwhile, ESLint (as recommended by their team) is not designed to be a code formatter!

## Prettier

[Prettier](https://prettier.io), a younger player became a superstar in the field of code formatting tools. It has amazing Editor support in addition to CLI, is widely used by over of 83% of developers according to [State of JS 2021 Report](https://2021.stateofjs.com/en-US/other-tools/#utilities), supports multi-languages, and most importantly, has few options!

Summary of [Prettier default style spec](https://prettier.io/docs/en/options.html#bracket-spacing):
- Print width of 80 columns
- Semicolons;
- Quotes for "strings" (smart switch to single: 'Like "this"')
- Trailing commas, (ES5)

If I want to describe the above spec, they are **safe**, smart, flexible, and also closer to other programming languages when you switch between them.

Prettier's IDE  support and speed are AWESOME but it turned out our friends working on StandardJS and Prettier had different opinions about code style. However with the few options, prettier provides it can be semi-compatible with StandardJS, I find customization against any definition of standard and a slippery area that introduces inconsistencies in the first place.

## My personal take

As a developer switching between repositories with different code styles, all I care about is that **it just formats while I type** and I loved the promises of prettier: Do not worry about code formatting. It just works! It automatically adds semicolons, wraps my code, smartly switches between quote and single quote, and takes care of everything besides my code.

The javascript ecosystem is already diverse enough. We can never define "the standard". Yet we can move towards a unified ecosystem by putting aside personal habits and preferences and agreeing on one opinion that is used and accepted most.

Prettier surprisingly kept me productive as I was before with previous code styles and automatically applies an optimal and safe code style which I trust had been well thought to set by bright minds. 

While knowing about prettier for a long time, it took me months to research and make a decision and finally apply it to a limited surface and try. During this time I looked at many popular codebases and they all made different decisions and customizations. I couldn't find "the one" answer. I decided to trust the authentic prettier choices. A noticeable number of popular packages use semicolons (including Node.js, Typeform, and Automatic/WordPress). [Deno](https://github.com/denoland/deno) became my main inspiration to finalize the choice, trusting the author of Node.js, not being wrong about making revolutionary changes, and trusting prettier defaults. [Bun](https://github.com/oven-sh/bun) is also _almost_ in the same boat.

Prettier is an amazing tool that I chose to adopt for the UnJS ecosystem. The default configuration is set to prettier standard defaults but conditions always depend on the project and the involved people working on the code on daily basis. and we can always change decisions thanks to the modern tooling that applies formatting changes almost seamlessly and automatically.

Choose the right tool, then relax and let your code editor worry about the rest. Trying not to be [Richard](https://www.youtube.com/watch?v=V7PLxL8jIl8). In the end, the only thing matters is that that you enjoyed coding and wrote a beautiful piece of work. Cheers 🥂

**Comment by luyees** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
